base/applications/mspaint/scrollbox.cpp

```diff
@@ -9,31 +9,90 @@
+ * 0  view < canvas
+ * 1  view < canvas + scroll
+ * 2  view >= canvas + scroll
+ *
+ * Matrix of scrollbar presence (VERTICAL,HORIZONTAL) given by
+ * vertical x horizontal scrollbar modes (view:whole report):
+ *
+ *           horizontal mode
+ *             |      0      |      1      |      2
+ * vertical ---+-------------+-------------+------------
+ *   mode    0 |  TRUE,TRUE  |  TRUE,TRUE  |  TRUE,FALSE
+ *          ---+-------------+-------------+------------
+ *           1 |  TRUE,TRUE  | FALSE,FALSE | FALSE,FALSE
+ *          ---+-------------+-------------+------------
+ *           2 | FALSE,TRUE  | FALSE,FALSE | FALSE,FALSE
+ */
+
+struct ScrollbarPresence
+{
+    BOOL bVert;
+    BOOL bHoriz;
+};
+
+CONST ScrollbarPresence sp_mx[3][3] =
+{
+    { {  TRUE,TRUE  }, {  TRUE,TRUE  }, {  TRUE,FALSE } },
+    { {  TRUE,TRUE  }, { FALSE,FALSE }, { FALSE,FALSE } },
+    { { FALSE,TRUE  }, { FALSE,FALSE }, { FALSE,FALSE } }
+};
+
 
 /* FUNCTIONS ********************************************************/
 
 void
 UpdateScrollbox()
 {
-    RECT clientRectScrollbox;
-    RECT clientRectImageArea;
+    CONST INT EXTRASIZE = 5; /* 3 px of selection markers + 2 px of border */
+
+    CRect tempRect;
+    CSize sizeImageArea;
+    CSize sizeScrollBox;
+    INT vmode, hmode;
     SCROLLINFO si;
-    scrollboxWindow.GetClientRect(&clientRectScrollbox);
-    imageArea.GetClientRect(&clientRectImageArea);
+
+    scrollboxWindow.GetWindowRect(&tempRect);
+    sizeScrollBox = CSize(tempRect.Width(), tempRect.Height());
+
+    imageArea.GetClientRect(&tempRect);
+    sizeImageArea = CSize(tempRect.Width(), tempRect.Height());
+    sizeImageArea += CSize(EXTRASIZE * 2, EXTRASIZE * 2);
+
+    /* show/hide the scrollbars */
+    vmode = (sizeScrollBox.cy < sizeImageArea.cy? 0 :
+                (sizeScrollBox.cy < sizeImageArea.cy +
+                    ::GetSystemMetrics(SM_CYHSCROLL) ? 1 : 2));
+    hmode = (sizeScrollBox.cx < sizeImageArea.cx ? 0 :
+                (sizeScrollBox.cx < sizeImageArea.cx +
+                    ::GetSystemMetrics(SM_CXVSCROLL) ? 1 : 2));
+    scrollboxWindow.ShowScrollBar(SB_VERT, sp_mx[vmode][hmode].bVert);
+    scrollboxWindow.ShowScrollBar(SB_HORZ, sp_mx[vmode][hmode].bHoriz);
+
     si.cbSize = sizeof(SCROLLINFO);
     si.fMask  = SIF_PAGE | SIF_RANGE;
-    si.nMax   = clientRectImageArea.right + 6 - 1;
     si.nMin   = 0;
-    si.nPage  = clientRectScrollbox.right;
+
+    si.nMax   = sizeImageArea.cx +
+                    (sp_mx[vmode][hmode].bVert == TRUE ?
+                        ::GetSystemMetrics(SM_CYHSCROLL) : 0);
+    si.nPage  = sizeScrollBox.cx;
     scrollboxWindow.SetScrollInfo(SB_HORZ, &si);
-    scrollboxWindow.GetClientRect(&clientRectScrollbox);
-    si.nMax   = clientRectImageArea.bottom + 6 - 1;
-    si.nPage  = clientRectScrollbox.bottom;
+
+    si.nMax   = sizeImageArea.cy +
+                    (sp_mx[vmode][hmode].bHoriz == TRUE ?
+                        ::GetSystemMetrics(SM_CXHSCROLL) : 0);
```

Comment:
I made a small mistake here. It should have been SM_CXVSCROLL instead of SM_CXHSCROLL.

Reply:
You can just make changes, amend it to the existing commit, and do force push.